File: app/src/main/java/de/bentolor/ahamanager/DatePickerFragment.java
package de.bentolor.ahamanager;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.view.View;
import android.widget.DatePicker;
import android.widget.DatePicker.OnDateChangedListener;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

public class DatePickerFragment extends DialogFragment {
    public static final String EXTRA_DATE = "de.bentolor.ahamoment.date";

    private Date mDate;

    public static DatePickerFragment newInstance(Date date) {
        Bundle args = new Bundle();
        args.putSerializable(EXTRA_DATE, date);

        DatePickerFragment fragment = new DatePickerFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        mDate = (Date) getArguments().getSerializable(EXTRA_DATE);

        // Create a calendar to get year,month,day
        final Calendar cal = Calendar.getInstance();
        cal.setTime(mDate);
        int year = cal.get(Calendar.YEAR);
        int month = cal.get(Calendar.MONTH);
        int day = cal.get(Calendar.DAY_OF_MONTH);

        View v = getActivity().getLayoutInflater()
                .inflate(R.layout.dialog_date, null);


        DatePicker datePicker = (DatePicker) v.findViewById(R.id.dialog_date_datePicker);
        datePicker.init(year, month, day, new OnDateChangedListener() {

            @Override
            public void onDateChanged(DatePicker view, int selectedYear, int monthOfYear,
                                      int dayOfMonth) {

                final Calendar calendar = Calendar.getInstance();
                calendar.setTime(mDate);
                int hour = calendar.get(Calendar.HOUR_OF_DAY);
                int minute = calendar.get(Calendar.MINUTE);
                // Translate picked date to Date
                mDate = new GregorianCalendar(selectedYear, monthOfYear, dayOfMonth, hour, minute)
                        .getTime();

                // Update argument to preserve selected value on rotation
                getArguments().putSerializable(EXTRA_DATE, mDate);
            }
        });

        return new AlertDialog.Builder(getActivity())
                .setView(v)
                .setTitle(R.string.date_picker_title)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        sendResult(Activity.RESULT_OK);
                    }
                })
                .create();
    }

    private void sendResult(int resultCode) {
        if (getTargetFragment() == null) {
            return;
        }

        Intent i = new Intent();
        i.putExtra(EXTRA_DATE, mDate);

        getTargetFragment()
                .onActivityResult(getTargetRequestCode(), resultCode, i);
    }
}

File: app/src/main/kotlin/de/bentolor/katamanager/DatePickerFragment.kt
package de.bentolor.katamanager

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.DialogFragment
import de.bentolor.ahamanager.R
import kotlinx.android.synthetic.main.dialog_date.*
import java.util.*

class DatePickerFragment : DialogFragment() {

    private var mDate: Date? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        mDate = arguments.getSerializable(EXTRA_DATE) as Date

        // Create a calendar to get year,month,day
        val cal = Calendar.getInstance()
        cal.time = mDate
        val year = cal.get(Calendar.YEAR)
        val month = cal.get(Calendar.MONTH)
        val day = cal.get(Calendar.DAY_OF_MONTH)

        val view = activity.layoutInflater.inflate(R.layout.dialog_date, null)

        dialog_date_datePicker.init(year, month, day) { view, y, m, dOM ->
            val calendar = Calendar.getInstance()
            calendar.time = mDate
            val hour = calendar.get(Calendar.HOUR_OF_DAY)
            val minute = calendar.get(Calendar.MINUTE)
            // Translate picked date to Date
            mDate = GregorianCalendar(y, m, dOM, hour, minute).time

            // Update argument to preserve selected value on rotation
            arguments.putSerializable(EXTRA_DATE, mDate)
        }

        return AlertDialog.Builder(activity)
                .setView(view)
                .setTitle(R.string.date_picker_title)
                .setPositiveButton(android.R.string.ok) { dialog, which -> sendResult(Activity.RESULT_OK) }
                .create()
    }

    private fun sendResult(resultCode: Int) {
        with(Intent()) {
            putExtra(EXTRA_DATE, mDate)
            targetFragment?.onActivityResult(targetRequestCode, resultCode, this)
        }
    }

    companion object {
        val EXTRA_DATE = "de.bentolor.ahamoment.date"

        fun newInstance(date: Date): DatePickerFragment {
            val args = Bundle()
            args.putSerializable(EXTRA_DATE, date)

            val fragment = DatePickerFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
